tensorstore/kvstore/s3/credentials/file_credential_provider.cc

```diff
@@ -0,0 +1,147 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorstore/kvstore/s3/credentials/file_credential_provider.h"
+
+#include <fstream>
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "absl/log/absl_log.h"
+#include "absl/status/status.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/str_cat.h"
+#include "absl/time/time.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/util/result.h"
+
+using ::tensorstore::internal::GetEnv;
+using ::tensorstore::internal::JoinPath;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// Credentials file environment variable
+static constexpr char kEnvAwsCredentialsFile[] = "AWS_SHARED_CREDENTIALS_FILE";
+// Default path to the AWS credentials file, relative to the home folder
+static constexpr char kDefaultAwsCredentialsFilePath[] = ".aws/credentials";
+// AWS user identifier
+static constexpr char kCfgAwsAccessKeyId[] = "aws_access_key_id";
+// AWS user password
+static constexpr char kCfgAwsSecretAccessKeyId[] = "aws_secret_access_key";
+// AWS session token
+static constexpr char kCfgAwsSessionToken[] = "aws_session_token";
+// Discover AWS profile in environment variables
+static constexpr char kEnvAwsProfile[] = "AWS_PROFILE";
+// Default profile
+static constexpr char kDefaultProfile[] = "default";
+
+Result<std::string> GetAwsCredentialsFileName() {
+  std::string result;
+
+  if (auto credentials_file = GetEnv(kEnvAwsCredentialsFile);
+      credentials_file) {
+    result = *credentials_file;
+  } else {
+    if (auto home_dir = GetEnv("HOME"); home_dir) {
+      result = JoinPath(*home_dir, kDefaultAwsCredentialsFilePath);
+    } else {
+      return absl::NotFoundError("Could not read $HOME");
+    }
+  }
+
+  if (auto fstream = std::ifstream(result.c_str()); !fstream.good()) {
```

Comment:
There's no need to lookup the filename here; an attempt is made to open it below, which will fail appropriately.

Reply:
removed